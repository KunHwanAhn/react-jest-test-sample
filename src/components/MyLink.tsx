import React, { useState, ReactNode } from 'react';

const STATUS = {
  HOVERED: 'hovered',
  NORMAL: 'normal',
} as const;

interface MyLinkProps {
  children: ReactNode;
  page?: string;
}
export default function MyLink({ children, page }: MyLinkProps) {
  const [status, setStatus] = useState<string>(STATUS.NORMAL);

  const updateHoveredStatus = () => {
    setStatus(() => STATUS.HOVERED);
  };

  const updateNormalStatus = () => {
    setStatus(() => STATUS.NORMAL);
  };

  return (
    <a
      className={status}
      href={page || '#'}
      onMouseEnter={updateHoveredStatus}
      onMouseLeave={updateNormalStatus}
    >
      {children}
    </a>
  );
}
MyLink.defaultProps = {
  page: undefined,
};
