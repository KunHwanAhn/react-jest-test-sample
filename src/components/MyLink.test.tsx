import React from 'react';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';
import MyLink from './MyLink';

const getTree = (result: null | ReactTestRendererJSON | ReactTestRendererJSON[]): ReactTestRendererJSON => {
  if (!result) {
    throw new Error('invalid tree');
  }

  let tree: ReactTestRendererJSON;
  if (Array.isArray(result)) {
    [tree] = result;
  } else {
    tree = result;
  }

  return tree;
};

describe('Link Component Test', () => {
  it('changes the class when hovered', () => {
    const component = renderer.create(
      <MyLink page="https://www.facebook.com">Facebook</MyLink>,
    );
    let tree = getTree(component.toJSON());

    expect(tree).toMatchSnapshot();

    // manually trigger the callback
    renderer.act(() => { tree.props.onMouseEnter(); });

    // re-rendering
    tree = getTree(component.toJSON());
    expect(true).toMatchSnapshot();

    // manually trigger the callback
    renderer.act(() => { tree.props.onMouseLeave(); });

    // re-rendering
    tree = getTree(component.toJSON());
    expect(true).toMatchSnapshot();
  });
});
