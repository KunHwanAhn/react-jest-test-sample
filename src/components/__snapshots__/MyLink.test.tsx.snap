// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link Component Test changes the class when hovered 1`] = `
<a
  className="normal"
  href="https://www.facebook.com"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Facebook
</a>
`;

exports[`Link Component Test changes the class when hovered 2`] = `true`;

exports[`Link Component Test changes the class when hovered 3`] = `true`;
